docs/terminals.md

```diff
@@ -0,0 +1,141 @@
+When running in detached runc mode, you can run your terminal either in Pass-Through or New-Terminal mode.
+
+You run in detached mode by doing one of the following:
+
+* `runc -d`
+* `runc create` followed by `runc start`, i.e. container lifecycle management
+
+#### Pass-Through Detached
+When running runc detached with terminal Pass-Through, the stdin/stdout/stderr of the calling process will be passed to `runc` and from there to the container's process. However, since `runc` detaches, there are several side effects to take into account:
+
+* The stdout/stderr of the process can become intermixed with the stdout/stderr of the calling process.
+* Any input into the calling process can cause indeterminate issues with which process actually receives the input.
+
+For example:
+
+```
+$ runc create foo              # container created, inheriting stdio from this terminal
+$ runc start foo               # container started, inheriting stdio from this terminal
+$ curl https://containerd.io   # get contents of this site to the terminal
+
+<!DOCTYPE html>
+I am container output
+<html lang="en">
+  <head>
+	  ASashcned more container output
+    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <title>containerd</title>
+```
+
+Two problems occurred above:
+
+1. The output from the `curl` command and the output of the container `foo` became intermingled.
+2. Our typing `curl https://containerd.io` actually was passed to both our shell and the container `foo`. The results can be unpredictable. Our example here was simulated and might not even work.
+
+Further, a third, possibly temporary problem exists. Pass-Through Detached only works correctly when stdout/stderr are either an actual terminal or a file. If you use _anything_ else, e.g. a pipe, it will hang, for reasons unknown. There is an [open issue](https://github.com/opencontainers/runc/issues/1721) on this.
+
+For this reason, we **strongly recommend** using New Terminal mode when running `runc` detached.
+
+#### New Terminal Detached
+When running runc detached with New Terminal mode, runc behaves exactly as described above, creating a new pty, passing it to the container, sending the fd to a socket, and exiting.
+
+## How To Get the FD for New Terminal
+When running `runc` in New Terminal mode, i.e. `terminal:true` in `config.json` - whether foreground or detached `runc` - you **must** tell `runc` what process will control the master end of the pty. To do that:
+
+1. Open a Unix-domain socket
+2. Pass the path to the socket to `runc create <container> --console-socket <path_to_socket>`
+3. Listen on the socket to get the master file descriptor
+4. Use that file descriptor to control stdio
+
+Note the one shortcoming: the single descriptor means that you have mixed stdout/stderr. There is no way to separate them. Essentially, this is a _console_, which has no separate stdout and stderr.
```

Comment:
(Though this shortcoming is identical to regular terminal emulators -- take a look at `/proc/self/fd/[12]` in your shell right now.)

Reply:
Of course. It really is only something you can separate when you start a new process with separate descriptors. Happy to add a comment here.